include/dali/dali.h

```diff
@@ -0,0 +1,1071 @@
+ *
+ * @warning When NOT using DALI_EXEC_IS_DYNAMIC, the maximum number of live daliPipelineOutputs_h
+ *          obtained from a single pipeline must not exceed the prefetch_queue_depth. An attempt
+ *          to run the pipeline again after the
+ */
+DALI_API daliResult_t daliPipelineOutputsDestroy(daliPipelineOutputs_h out);
+
+typedef struct _DALIOperatorTrace {
+  const char *operator_name;
+  const char *trace;
+  const char *value;
+} daliOperatorTrace_t;
+
+/** Gets all operator "traces" that were set when producing this set of outputs.
+ *
+ * @param outputs         [in]  The outputs
+ * @param out_traces      [out] A return value pointer where, the a pointer to the beginning of an
+ *                              array of operator traces is stored.
+ * @param out_trace_count [out] A pointer that receives the number of traces.
+ *
+ * The output array is valid until the `outputs` handle is destroyed.
+ */
+DALI_API daliResult_t daliPipelineOutputsGetTraces(
+  daliPipelineOutputs_h outputs,
+  const daliOperatorTrace_t **out_traces,
+  int *out_trace_count);
+
+/** Gets a single operator "trace", identified by operator instance name and a trace name.
+ *
+ * @param outputs       [in]  The outputs
+ * @param out_trace     [out] A pointer which receives a ppointer to the trace.
+ * @param operator_name [in]  The name of the operator whose trace is being obtained.
+ * @param trace_name    [in]  The name of the trace.
+ *
+ * @retval DALI_SUCCESS           On success
+ * @retval DALI_ERROR_INVALID_KEY When there's no trace that matches the names
+ */
+DALI_API daliResult_t daliPipelineOutputsGetTrace(
+  daliPipelineOutputs_h outputs,
+  const char **out_trace,
+  const char *operator_name,
+  const char *trace_name);
+
+/** Gets index-th output.
+ *
+ * The handle returned by this function must be released with a call to daliTensorListDecRef.
+ *
+ * Unless the pipeline uses DALI_EXEC_IS_DYNAMIC flag, the returned tensor list must not be used
+ * after the `outputs` handle is destroyed.
+ *
+ * @param outputs [in]  The pipeline outputs object
+ * @param out     [out] A pointer to a TensorList handle
+ * @param index   [in]  The index of the output to get a handle to.
+ */
+DALI_API daliResult_t daliPipelineOutputsGet(
+  daliPipelineOutputs_h outputs,
+  daliTensorList_h *out,
+  int index);
+
+/****************************************************************************/
+/*** Tensor and TensorList API **********************************************/
+/****************************************************************************/
+
+typedef struct _DALITensorDesc {
+  /** The number of dimensions of the tensor.
+   *
+   * 0 denotes a scalar value. Negative values are invalid.
+   */
+  int             ndim;
+
+  /** The shape of the tensor.
+   *
+   * The shape starts with the "slowest" dimension - a row-major 640x480 interleaved RGB image
+   * would have the shape [480, 640, 3].
+   *
+   * The shape can be NULL if ndim == 0
```

Comment:
Can or is? Should we limit ambiguity?

Reply:
Definitely "can". Overvevaluation is evil. If you have `std::vector<int>` with a nonzero capacity but zero size, then `.data()` will return a non-null pointer. We don't want to force the user to "fix" std::vector - that'd be daft.